src/core/system.cpp

```diff
@@ -2320,4 +2298,71 @@ uint8_t System::systemStatus() {
     return systemStatus_;
 }
 
+// take a string range like "6-11, 1, 23, 24-48" which has optional ranges and single values convert to a vector of ints
+std::vector<uint8_t> System::string_range_to_vector(const std::string & range) {
+    std::vector<uint8_t>   valid_gpios;
+    std::string::size_type pos  = 0;
+    std::string::size_type prev = 0;
+    while ((pos = range.find(',', prev)) != std::string::npos) {
+        valid_gpios.push_back(std::stoi(range.substr(prev, pos - prev)));
+        prev = pos + 1;
+    }
+    valid_gpios.push_back(std::stoi(range.substr(prev)));
+    return valid_gpios;
+}
+
+// return the list of valid GPIOs
+// note: we do not allow 0, which is used sometimes to indicate a disabled pin
+std::vector<uint8_t> System::valid_gpio_list() {
+    // get free gpios based on board/platform type
+#if CONFIG_IDF_TARGET_ESP32C3
+    // https://www.wemos.cc/en/latest/c3/c3_mini.html
+    std::vector<uint8_t> valid_gpios = string_range_to_vector("11-19, 21"); // can go higher than 21 on some boards
+#elif CONFIG_IDF_TARGET_ESP32S2
+    std::vector<uint8_t> valid_gpios = string_range_to_vector("19-20, 22-32, 40");
+#elif CONFIG_IDF_TARGET_ESP32S3
+    std::vector<uint8_t> valid_gpios = string_range_to_vector("19-20, 22-37, 39-42, 48");
+#elif CONFIG_IDF_TARGET_ESP32 || defined(EMSESP_STANDALONE)
+    std::vector<uint8_t> valid_gpios = string_range_to_vector("1, 6-11, 16-17, 20, 24, 28-31, 40");
+#else
+    std::vector<uint8_t> valid_gpios = {};
+#endif
+
```

Comment:
Look like you have listed invalid pins as valid. I've checked the datasheet: Unclear, list strapping pins (esp32 0,2,5)? How to manage input only pins for esp32: 34-39
ESP32 valid: 0, 2, 4, 5,12-15, 18, 19, 23, 25-27 ,32-39
S3 valid: 2, 4-14, 17, 18, 21, (33-38), 45, 46
33-38 are used for octal SPI for 32M vchip version
S2 valid: 0-14, 19, 20, 21 ,33-38, 45,46
C3 valid: 0-10

Reply:
Thanks for reviewing. I based the valid pins logic on what we already had in code, specifically in `validators.ts` and `System::is_valid_gpio()`, without making any changes.

Pin 0 is a special case because we sometimes use this to indicate disabled, as in Dallas and LED

I would block the GPIO input only pins for now. If someone needs it we can always make an exception. 

Shall I adjust based on what you researched, or commit and you make the changes?

Reply:
done.